### ISSUE TITLE
Conformance with commonmark

### ISSUE THREAD

**Issue by tombentley** [rohitmohan96/ceylon.markdown#12]
We're not currently conforming with the commonmark spec. 

I've pushed a branch (`spectests`) to my fork which adds all the tests used by the commonmark reference implementation. There are a lot of tests (>600), and there are a lot which fail. I don't know how many because many get stuck due to infinite loops.

Obviously before we can merge the tests we need to fix the implementation.

**Comment by rohitmohan96** [comment]
At a glance, some of the errors seem to be because the html generated by `ceylon.html` isn't exactly same as the one in the spec (with respect to indentations, breaks, etc.) but some tests do generate the wrong output. I don't think I have had any stuck in infinite loops.

**Comment by tombentley** [comment]
Yeah, after I opened this issue I did notice a few things like `<hr>` which
seems to be `ceylon.html`s fault, so we should open bugs for those.

**Comment by jvasileff** [comment]
`hr` elements are void and have no corresponding end tags, if that's the issue. See https://www.w3.org/TR/html5/syntax.html#void-elements

**Comment by rohitmohan96** [comment]
The problem is that the html generated by `ceylon.html` isn't identical to the one in the spec.  
Example, 
In the spec, 

```
<ul>
<li>foo</li>
<li></li>
<li>bar</li>
</ul>
```

and this is generated by `ceylon.html`: 

```
<ul>
  <li>foo</li>
  <li></li>
  <li>bar</li>
</ul>
```

The only difference is the indentation which is causing these tests to fail. 

Similarly, the `<hr />` and `<br />` in the spec is generated as `<hr>` and `<br>`.

**Comment by tombentley** [comment]
Well maybe `ceylon.html::RenderingConfiguration` needs to support pretty printing without indentation.

I thought there was an XML compatible way to serialize HTML5, in which case it would be good is `ceylon.html` could support that too (since non-XML is such a pain to parse unless you're a browser). @thradec is probably the man to ask about these things.

**Comment by thradec** [comment]
> Well maybe ceylon.html::RenderingConfiguration needs to support pretty printing without indentation.

we can add such parameter into `RenderingConfiguration` and use it [here](https://github.com/ceylon/ceylon-sdk/blob/master/source/ceylon/html/internal/HtmlRenderer.ceylon#L174), instead of hardcoded 2

> @thradec is probably the man to ask about these things.

@jvasileff wrote most of the rendering logic

**Comment by tombentley** [comment]
> we can add such parameter

OK, done.

**Comment by jvasileff** [comment]
> Similarly, the `<hr />` and `<br />` in the spec is generated as `<hr>` and `<br>`. 

I added an option `closeVoidElements` which can be used to generate `<hr />` as the tests expect.

**Comment by tombentley** [comment]
Great, thanks @jvasileff!

@rohitmohan96 do you need a `ceylon.html/1.3.0-1` released so that you can use these new options?

**Comment by rohitmohan96** [comment]
Thanks @tombentley and @jvasileff . That fixed some of the tests. 

Another issue. This is what the spec has: 

``` html
<p><a href="http://foo.bar.baz">http://foo.bar.baz</a></p>
```

This is what `ceylon.html` generates: 

``` html
<p>
<a href="http://foo.bar.baz">http://foo.bar.baz</a>
</p>
```

> @rohitmohan96 do you need a ceylon.html/1.3.0-1 released so that you can use these new options?

The project uses `1.3.1` as of now. So I already have the new options.

**Comment by jvasileff** [comment]
Is `<p>` the only one left, or are there other rendering differences?

It _might_ work to have a `[String*] dontIndentElements` setting that would be used in:

``` ceylon
                previousStartOrEndTagWasBlock = prettyPrint
                        && indentElements.contains(elementName.lowercased);
```

but we're also getting pretty specific with these formatting options. @tombentley WDYT?

**Comment by tombentley** [comment]
I'm not really sure that would be justifiable for `ceylon.html`. 

I can't see where in the HTML5 spec it says precisely when it's OK to ignore whitespace between elements (and therefore, when it would & would not be OK for `ceylon.html` to add whitespace as part of serialization). Whether those rules would be compatible with what the conformance tests expect isn't clear either.

So maybe the answer is for `test.ceylon.markdown` to use its own HTML serialization, which can take these arbitrary-seeming choices into account.

**Comment by jvasileff** [comment]
I thought we might be able to change the tests instead of the serializer, but I see that's impractical due to the number of tests.

> So maybe the answer is for test.ceylon.markdown to use its own HTML serialization, which can take these arbitrary-seeming choices into account

I'm thinking the feature bloat on the serializer would be better than forgoing integrated tests, especially considering potential escaping issues.

@rohitmohan96 can you answer:

> Is `<p>` the only one left, or are there other rendering differences?

**Comment by rohitmohan96** [comment]
> Is <p> the only one left, or are there other rendering differences?

Well, there are similar ones but it is difficult to go through all the tests since there are so many. 

I guess I could write my own visitor that could generate html similar to the spec. 

> I thought we might be able to change the tests instead of the serializer, but I see that's impractical due to the number of tests.

This would also cause problems each time the spec is updated since they have to be changed every time.

**Comment by CPColin** [comment]
I commented on jgm/CommonMark#274 to see if implementors could target AST compliance instead of an exact HTML rendering. In the meantime, I forked @tombentley's spectests branch, merged @rohitmohan96's latest changes, and pushed a few changes of my own here:

https://github.com/CPColin/ceylon.markdown/tree/spectests

The `testSpec` function includes a few hacks to reduce the amount of noise, making it easier to see which failures are probably `ceylon.markdown`'s parsing, probably the spec's HTML rendering, or probably `ceylon.html`'s rendering.

For example, from the 221 failing tests:

- [ ] `specTest_Autolinks_572`, `specTest_Backslash_escapes_295`, `specTest_Code_spans_325`: `ceylon.html` should be percent-encoding backslashes, brackets, and backticks.
- [ ] `specTest_Backslash_escapes_297`, `specTest_Backslash_escapes_298`: `ceylon.html` renders attributes in alphabetical order. Either a custom renderer would have to flip the attributes back around or we'd need to target AST compliance here.
- [ ] `specTest_Backslash_escapes_299`: `ceylon.markdown` is not preserving the code language.
- [ ] `specTest_Block_quotes_(various)`: `blockquote` elements tend to end too late. Some of the tests expect a line break in elements that `ceylon.html` renders as empty.
- [ ] `specTest_Emphasis_and_strong_emphasis_344`: `ceylon.markdown` is starting a list, when `CommonMark` only starts a list when the `*` is followed by a space.
- [ ] `specTest_Lists_263`: The change in delimiter should start a new list.
- [ ] `specTest_Entity_and_numeric_character_references_(various)`: I don't know why the spec appears to mandate _decoding_ HTML entities. `ceylon.html` is dutifully encoding the ampersands.

It looks like there are plenty of tests that are failing because of renderer differences, but also a good amount that are legitimate.

**Comment by jvasileff** [comment]
For: 

> specTest_Backslash_escapes_298: ceylon.html renders attributes in alphabetical order. Either a custom renderer would have to flip the attributes back around or we'd need to target AST compliance here.

The code is specifying attributes using attribute-specific arguments to the `A` constructor:

```ceylon
        shared actual default HtmlNode visitLink(Link link) => A {
                href = link.destination;
                title = if (link.title == "") then null else link.title;
                children = [for (child in link.children)
                                if (is FlowCategory|String ch = child.accept(this)) ch];
        };
```
but if the attributes were provided to the more generic `Attributes attributes` parameter, it's possible the order would be preserved. See https://modules.ceylon-lang.org/repo/1/ceylon/html/1.3.2/module-doc/api/A.type.html

**Comment by CPColin** [comment]
Good call. I made that change in both `visitImage` and `visitLink` and we're down to 192 failures.

**Comment by CPColin** [comment]
Those 192 failures are when running on the JVM backend; running on the JS backend produces 257 failures, not all of them the same, somehow.

**Comment by rohitmohan96** [comment]
Could be related to #4.

**Comment by CPColin** [comment]
It gets even weirder: three of the tests that fail on the JVM backend pass on the JS backend. (Test 388, 391, and 403.)

I just pushed a big update to the `spectests` branch on my repo. It's a different take on what Tom Bentley did above that brings in the `commonmark.js` module so we can compare the AST's directly. You can see some of them differing by line breaks, similar to how the generated HTML can differ. When both the generated HTML _and_ the generated AST are wrong, it's a likely indication that the parser has an issue. The downside is you have to run the tests (in the `test.ceylon.markdown.spectests` package) on the JS backend, and that can take a while. Eclipse will let you compare those results against the results when run on the JVM, if you want to see the differences.
